plugin/src/com/microsoft/alm/plugin/idea/tfvc/core/ReactiveTfvcClient.java

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root.
+
+package com.microsoft.alm.plugin.idea.tfvc.core;
+
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.project.Project;
+import com.microsoft.alm.plugin.authentication.AuthenticationInfo;
+import com.microsoft.alm.plugin.context.ServerContext;
+import com.microsoft.alm.plugin.external.models.PendingChange;
+import com.microsoft.alm.plugin.external.reactive.ReactiveTfvcClientHolder;
+import org.jetbrains.annotations.NotNull;
+
+import java.net.URI;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Stream;
+
+/**
+ * The new, "reactive" TFVC client implementation. Most of the methods are asynchronous and delegate to the external
+ * client that is always running. Works with Rd protocol.
+ */
+public class ReactiveTfvcClient implements TfvcClient {
+
+    private static Logger ourLogger = Logger.getInstance(ReactiveTfvcClient.class);
+
+    @NotNull
+    private final Project myProject;
+
+    public ReactiveTfvcClient(@NotNull Project project) {
+        myProject = project;
+    }
+
+    @Override
+    @NotNull
+    public CompletableFuture<List<PendingChange>> getStatusForFilesAsync(
+            @NotNull ServerContext serverContext,
+            @NotNull List<String> pathsToProcess) {
+        long startTime = System.nanoTime();
+
+        URI collectionUri = serverContext.getCollectionURI();
+        AuthenticationInfo authenticationInfo = serverContext.getAuthenticationInfo();
+        Stream<Path> paths = pathsToProcess.stream().map(Paths::get);
+
+        return ReactiveTfvcClientHolder.getInstance(myProject).getClient()
```

Comment:
The responsibility chain is like so:
1. Some external code calls `TfvcClient.getInstance()`
2. Then, `TfvcClient` calls its appropriate implementation (either `Classic` or `Reactive`)
3. Then, the implementation does the stuff _knowing it has been chosen over the alternatives_

So, in general, implementation-specific code (like `ReactiveTfvcClient`) shouldn't call the general code (like `TfvcClient`), just because it works the other way  around.

In this particular place, it wouldn't compile: `ReactiveTfvcClientHolder.getInstance(myProject).getClient()` returns a `ReactiveTfvcClientHost` which performs all the low-level operations such `tf status`, `tf whatever` _for a particular TFVC workspace_, and `TfvcClient` is a high-level entity that has no idea of TFVC workspaces.